src/Storages/Kafka/StorageKafka2.cpp

```diff
@@ -0,0 +1,1490 @@
+        kafka_settings->input_format_allow_errors_num = 0;
+        kafka_settings->input_format_allow_errors_ratio = 0;
+    }
+    StorageInMemoryMetadata storage_metadata;
+    storage_metadata.setColumns(columns_);
+    setInMemoryMetadata(storage_metadata);
+    setVirtuals(createVirtuals(kafka_settings->kafka_handle_error_mode));
+
+    auto task_count = thread_per_consumer ? num_consumers : 1;
+    for (size_t i = 0; i < task_count; ++i)
+    {
+        auto task = getContext()->getMessageBrokerSchedulePool().createTask(log->name(), [this, i] { threadFunc(i); });
+        task->deactivate();
+        tasks.emplace_back(std::make_shared<TaskContext>(std::move(task)));
+    }
+
+    const auto first_replica = createTableIfNotExists();
+
+    if (!first_replica)
+        createReplica();
+
+    activating_task = getContext()->getSchedulePool().createTask(log->name() + "(activating task)", [this]() { activate(); });
+    activating_task->deactivate();
+}
+
+VirtualColumnsDescription StorageKafka2::createVirtuals(StreamingHandleErrorMode handle_error_mode)
+{
+    VirtualColumnsDescription desc;
+
+    desc.addEphemeral("_topic", std::make_shared<DataTypeLowCardinality>(std::make_shared<DataTypeString>()), "");
+    desc.addEphemeral("_key", std::make_shared<DataTypeString>(), "");
+    desc.addEphemeral("_offset", std::make_shared<DataTypeUInt64>(), "");
+    desc.addEphemeral("_partition", std::make_shared<DataTypeUInt64>(), "");
+    desc.addEphemeral("_timestamp", std::make_shared<DataTypeNullable>(std::make_shared<DataTypeDateTime>()), "");
+    desc.addEphemeral("_timestamp_ms", std::make_shared<DataTypeNullable>(std::make_shared<DataTypeDateTime64>(3)), "");
+    desc.addEphemeral("_headers.name", std::make_shared<DataTypeArray>(std::make_shared<DataTypeString>()), "");
+    desc.addEphemeral("_headers.value", std::make_shared<DataTypeArray>(std::make_shared<DataTypeString>()), "");
+
+    if (handle_error_mode == StreamingHandleErrorMode::STREAM)
+    {
+        desc.addEphemeral("_raw_message", std::make_shared<DataTypeString>(), "");
+        desc.addEphemeral("_error", std::make_shared<DataTypeString>(), "");
+    }
+
+    return desc;
+}
+void StorageKafka2::partialShutdown()
+{
+    for (auto & task : tasks)
+    {
+        LOG_TRACE(log, "Cancelling streams");
+        task->stream_cancelled = true;
+    }
+
+    for (auto & task : tasks)
+    {
+        LOG_TRACE(log, "Waiting for cleanup");
+        task->holder->deactivate();
+    }
+    is_active = false;
+}
+
+bool StorageKafka2::activate()
+{
+    LOG_TEST(log, "activate task");
+    if (is_active && !getZooKeeper()->expired())
+    {
+        LOG_TEST(log, "No need to activate");
+        return true;
+    }
+
+    if (first_time)
+    {
+        LOG_DEBUG(log, "Activating replica");
+        assert(!is_active);
+    }
+    else if (!is_active)
+    {
+        LOG_WARNING(log, "Table was not active. Will try to activate it");
+    }
+    else if (getZooKeeper()->expired())
```

Comment:
Maybe I misunderstood something, but if Zookeeper expires between line 206 and here, that means:
1. ZK was alive on line 206, so `is_active` had to be `false` to not return on line 209.
2. if `is_active` is `false` on line 206, then we run into line 219, thus avoiding `UNREACHABLE`.

Otherwise, if ZK is already expired on line 206, we run into this condition, thus avoiding `UNREACHABLE`.

I don't see any issues with this logic, but it is possible that I miss some important detail, so please don't hesitate to correct me.